File: src/main/java/graphql/execution/SimpleExecutionStrategy.java
package graphql.execution;

import graphql.ExecutionResult;
import graphql.ExecutionResultImpl;
import graphql.language.Field;
import graphql.schema.GraphQLObjectType;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public class SimpleExecutionStrategy
    extends ExecutionStrategy
{
    @Override
    public CompletionStage<ExecutionResult> execute(ExecutionContext executionContext,
                                                    GraphQLObjectType parentType,
                                                    Object source,
                                                    Map<String, List<Field>> fields)
    {
        CompletableFuture<ExecutionResult> promise = new CompletableFuture<>();

        List<CompletableFuture> fieldPromises = new ArrayList<>();
        Map<String, Object> results = new LinkedHashMap<String, Object>();
        for (String fieldName : fields.keySet()) {
            List<Field> fieldList = fields.get(fieldName);
            fieldPromises.add((CompletableFuture)resolveField(executionContext, parentType, source, fieldList)
                .thenAccept(resolvedResult -> {
                    results.put(fieldName, resolvedResult != null ? resolvedResult.getData() : null);
                })
                .exceptionally(e -> {
                    promise.completeExceptionally(e);
                    return null;
                }));

        }

        CompletableFuture[] fieldPromisesArray = new CompletableFuture[fieldPromises.size()];
        CompletableFuture.allOf(fieldPromises.toArray(fieldPromisesArray))
                         .thenAccept(aVoid -> promise.complete(new ExecutionResultImpl(results, executionContext.getErrors())));
        return promise;
    }
}

File: src/main/kotlin/graphql/execution/SimpleExecutionStrategy.kt
package graphql.execution

import graphql.ExecutionResult
import graphql.ExecutionResultImpl
import graphql.language.Field
import graphql.schema.GraphQLObjectType
import graphql.util.awaitAll
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage

open class SimpleExecutionStrategy : AbstractExecutionStrategy() {
    override fun execute(executionContext: ExecutionContext,
                         parentType: GraphQLObjectType,
                         source: Any,
                         fields: Map<String, List<Field>>): CompletionStage<ExecutionResult> {
        val promise = CompletableFuture<ExecutionResult>()

        //println("**** Executing ${parentType.name} source ${source.javaClass.name} fields ${fields.keys.map { it }}")

        val results = LinkedHashMap<String, Any?>()
        fields.map { (fieldName, fieldList) ->
            resolveField(executionContext, parentType, source, fieldList)
                    .whenComplete { resolvedResult, ex ->
                        if (ex == null) {
                            results.put(fieldName, resolvedResult?.data())
                        } else
                            promise.completeExceptionally(ex)
                    }.toCompletableFuture()

        }.awaitAll().thenAccept {
            promise.complete(ExecutionResultImpl(results, executionContext.errors()))
        }

        return promise
    }
}

